### DISCUSSION TITLE
Location of settings

### DISCUSSION THREAD

**Discussion by Zoey2936** [nextcloud/all-in-one#1321] [Questions] [answered]
A Question I have a long time and I still don't understand is, why can some setting be set using the AIO-Panel and some needs to be set as env when starting the mastercontainer. Because some settings (not all!) which need to be set on mastercontainer startup, don't change the mastercontainer, only settings on the containers of the sub services.

**Comment by szaimen** [comment]
Because the config options will get written to the configuration.json by doing so, which makes the configuration easier as you do not need to make edits to the configuration.json file. Also by doing so, they are also restored upon restoring the whole instance since they are then directly written to the configuration.json. 

In principle you could also directly change the values in the configuration.json but we do not want to expose and document these internally used values and do not want to risk people destroying their instance by making a wrong edit.

Also not all values are available through the UI since that would make it very crowded.

I hope that makes sense!

**Reply by Zoey2936** [reply]
I don't understand the first two paragraphs, since the NEXTCLOUD_MOUNT is written into the configuration.json, which is set as env, as well as the backup location which can be set using the mastercontainer, what is the different that these two option need to be set on different places? Why can't both be set from the AIO interface?

The only reason I see in your answer is, that over this way, the UI looks cleaner.

I would understand it if the reason would be that the env settings also influence system settings like ports and system directories and the AIO panel only AIO internal settings, so that a Nextcloud admin don't have access to system settings through the aio panel and only nextcloud settings, but this seems not to be the case.

**Reply by szaimen** [reply]
> The only reason I see in your answer is, that over this way, the UI looks cleaner.

Indeed, this is the main point as in my opinion, the UI is already too crowded with options.

I would also rather not have introduced the NEXTCLOUD_MOUNT option apart from others but many people wanted this option so I added it but will definitely not make this available through the UI. It shall remain an advanced option for the people that need it.

**Reply by szaimen** [reply]
BTW, the maintainer of NextcloudPi currently thinks about integrating AIO into their project IIRC: https://github.com/nextcloud/nextcloudpi/issues/1589#issuecomment-1264593765

So adding a UI option for this would possibly be something for them if they integrate AIO at some point.

**Reply by Zoey2936** [reply]
> > The only reason I see in your answer is, that over this way, the UI looks cleaner.
> 
> Indeed, this is the main point as in my opinion, the UI is already too crowded with options.

I understand this, but would it then not be better that the UI gets optimized? Since in my opinion, it's not very user-friendly to have two places to change settings, while one place requires a full recreation of the mastercontainer, which is the most time only possible if the user has root access over ssh. (I know that these settings get not changed every day...)
I know that is very exaggerated now, but then theoretically all settings could be moved to env and then the AIO panel removed, if you understand what I mean with this.
And as said today in another Issue, some settings on the UI are hidden, while the containers are running and other settings are visible and blocked.

**Reply by szaimen** [reply]
> I understand this, but would it then not be better that the UI gets optimized?

Having all settings available through the UI will not optimize the UI.

> Since in my opinion, it's not very user-friendly to have two places to change settings, while one place requires a full recreation of the mastercontainer, which is the most time only possible if the user has root access over ssh.

This is a non-argument since AIO requires SSH root access anyway. Also, there is the same thing for Nextcloud. Some settings you can change via UI, some are advanced options and need to get changed via occ or in your config file - as I said changing them in the config file would be possible too but I don't want to expose them to the user but feel free to dig into the code and create a guide on that.

BTW, for our deployments, we even set some of the arguments by default which does not require to recreate the mastercontainer then:
https://github.com/nextcloud-releases/all-in-one/blob/a5da198cfbc5f4bc2bb6002d0adcaf29cd985321/.build/ova/build.sh#L19

> I know that is very exaggerated now, but then theoretically all settings could be moved to env and then the AIO panel removed, if you understand what I mean with this.

This is also something what I don't want to do. Either options get added via UI or via env for better maintainability. However changing a specific value in the configuration.json works here as well.

> And as said today in another Issue, some settings on the UI are hidden, while the containers are running and other settings are visible and blocked.

Yes, this is also indeed intentional to keep the UI as clean as possible. The spellcheck option for collabora was only added to the UI since for our deployments I indeed don't want to ask them to recreate the container so it was added but is mostly hidden.

**Reply by Zoey2936** [reply]
> > I understand this, but would it then not be better that the UI gets optimized?
> 
> Having all settings available through the UI will not optimize the UI.

With this I meant for example having Tabs on the AIO Panel, for example one for backup, one for ports and mounts and one for other settings and one for quick settings (start/stop/backup creation). Or something like this.

My main point/question is simply, that I don't know why setting have been places, where they are. The Timezone option was added to the UI on 23.5.2022, while the Nextcloud mount option was added as env earlier.

But the project is maintained by you and I have no clue of PHP, so I will respect your opinion.

I will try to write something for the configuration.json, when #1325 is published to develop.

But to something else - I think it would be a good Idea to add a Table of Contents to the Readme.md, what do you think? So a user can see, what option are possible to be set using env (so the user knows them directly before the first deploy of AIO) and so the user has a better overview of the Readme.md, without reading the compose file (for the env options) or the entire Readme.md?
If you think this is a good Idea, I could do this.

**Reply by szaimen** [reply]
> I will try to write something for the configuration.json, when #1325 is published to develop.

All right, cool!

> But to something else - I think it would be a good Idea to add a Table of Contents to the Readme.md, what do you think?

Having such a table sounds good to me! However can you create that in the wiki article that you are going to create? We could then add a link from the readme to this article! 

E.g: 

---

Valid for version 2.0.4:

| config description | env name | gui place | config name | link to instructions |
|---|---|---|---|---|
| Changing Nextclouds Datadir on the host | NEXTCLOUD_DATADIR | - | nextcloud_datadir | https://github.com/nextcloud/all-in-one#how-to-change-the-default-location-of-nextclouds-datadir |

---

WDYT?

**Reply by szaimen** [reply]
> I will try to write something for the configuration.json, when https://github.com/nextcloud/all-in-one/pull/1325 is published to develop.

This is now released with v2.1.0 Beta. Testing and feedback is welcome!

**Comment by szaimen** [comment, marked as answer]
So my advice for you would be creating a guide in https://github.com/nextcloud/all-in-one/discussions/new?category=wiki on how to adjust some values in the configuration.json directly but apart from that I do not see the improvement demand.
